### ISSUE TITLE
Unexpected typing error when trying to call a generic function that applies a generic function to a value

### ISSUE THREAD

**Issue by ericstalbot** [microsoft/pyright#2070]
**Describe the bug**

I'm seeing an unexpected typing error when trying to call a function that applies a generic function to a value.

**To Reproduce**

See the first example code below. This example code produces an unexpected typing error.

**Expected behavior**

I expected pyright would figure out that because the type of the second argument to `b` is `A[int]`,
the type of the first argument must something that satisifies `Callable[[A[int]], U]`, and that `f` would satisfy 
that requirement, without having to do the workaround (see second example below).

**Screenshots or Code**

```python
# this example has an unexpected typing error

from typing import TypeVar, Callable, Generic, Type


T = TypeVar('T')
U = TypeVar('U')


def b(func: Callable[[T], U], value: T) -> U:
    return func(value)


class A(Generic[U]):
    def __init__(self, value: U):
        self.value = value

def f(x: A[U]) -> U:
    return x.value


b(f, A(1))
# Argument of type "A[int]" cannot be assigned to parameter "value" of type "T@b" in function "b"
#  Type "A[U@f]" cannot be assigned to type "A[int]"PylancereportGeneralTypeIssues
```

```python
# this example is a work-around that doesn't have a typing error

from typing import TypeVar, Callable, Generic, Type


T = TypeVar('T')
U = TypeVar('U')


def b(func: Callable[[T], U], value: T) -> U:
    return func(value)


class A(Generic[U]):
    def __init__(self, value: U):
        self.value = value

def f(x: A[U]) -> U:
    return x.value


def g(t: Type[A[T]]) -> Callable[[A[T]], T]:
    return f

b(g(A[int]), A(1))
# ok
```

**VS Code extension or command-line**

- running pyright as a vs code extension via pylance
- pylance version 2021.7.2
- pyright version 1.1.155

**Additional context**

This seems to be a variation on the problem reported here: https://stackoverflow.com/questions/68132296/issue-with-using-a-typevar-as-an-argument-to-a-callable

Mypy also reports an error on the first example above: `Argument 1 to "b" has incompatible type "Callable[[A[U]], U]"; expected "Callable[[A[int]], U]"`. And mypy also does not report an error on the second example above. So either mypy and pyright have the same bug, or this is not a bug. 

Apologies in advance if this is a spurious bug report. I don't feel super qualified to report a bug here, but I'm also 
trying to learn how to do typing in python. If this is not a bug, any tips on how to fill my knowledge gap would be appreciated.

**Comment by erictraut** [comment]
The constraint solver in Python type checkers solve the type variables within a specified scope. In the example above, it is solving the type variables in the scope defined by function `b` (`T@b` and `U@b`). The type variable `U@f` is not in scope for `b`, so it is not solved as part of a call to `b`.

Your second example works because you are first solving for the type variables in a call to `g` and then passing this concrete return type as an argument to `b`.

So this is intended behavior, not a bug. 

I understand why you'd think this would work, but the Python type system (as defined in PEP 484, etc.) isn't able to accommodate this level of complexity.

**Comment by ericstalbot** [comment]
Thanks for the explanation!

**Comment by erictraut** [comment]
I've enhanced pyright's constraint solver so it can handle this case. This will be included in the next release.

**Comment by ericstalbot** [comment]
Nice!

**Comment by erictraut** [comment]
This is included in pyright 1.1.286, which I just published. It will also be included in a future version of pylance.
